geode-assembly/src/upgradeTest/java/org/apache/geode/management/RollingUpgradeWithSslDUnitTest.java

```diff
@@ -0,0 +1,276 @@
+    generateStores();
+    /*
+     * We must use absolute paths for truststore and keystore in properties file and
+     * since we don't know those at coding-time, we must generate the file.
+     * Since GfshRule provides no way to pass along Properties object to start server etc,
+     * we must write the properties to an actual file.
+     */
+    final Properties properties = generateSslProperties();
+
+    securityPropertiesFile = tempFolder.newFile("gfsecurity.properties");
+    final FileOutputStream fileOutputStream =
+        new FileOutputStream(securityPropertiesFile.getAbsolutePath());
+    properties.store(fileOutputStream, "");
+  }
+
+  @Test
+  public void testRollingUpgradeWithDeployment() throws Exception {
+    final int locatorPort = portSupplier.getAvailablePort();
+    final int locatorJmxPort = portSupplier.getAvailablePort();
+    final int locator2Port = portSupplier.getAvailablePort();
+    final int locator2JmxPort = portSupplier.getAvailablePort();
+    final int server1Port = portSupplier.getAvailablePort();
+    final int server2Port = portSupplier.getAvailablePort();
+
+    final GfshExecution startupExecution =
+        GfshScript.of(
+            startLocatorCommandWithConfig("locator1", locatorPort, locatorJmxPort, -1))
+            .and(startLocatorCommandWithConfig("locator2", locator2Port, locator2JmxPort,
+                locatorPort))
+            .and(startServerCommandWithConfig("server1", server1Port, locatorPort))
+            .and(startServerCommandWithConfig("server2", server2Port, locatorPort))
+            .execute(oldGfsh, tempFolder.getRoot());
+
+    initializeRegion(locatorPort);
+    causeP2PTraffic(locatorPort);
+
+    // doing rolling upgrades
+    upgradeLocator("locator1", locatorPort, locatorJmxPort, locator2Port, startupExecution);
+    verifyListMembers(locatorPort);
+    causeP2PTraffic(locatorPort);
+
+    upgradeLocator("locator2", locator2Port, locator2JmxPort, locatorPort, startupExecution);
+    verifyListMembers(locatorPort);
+    causeP2PTraffic(locatorPort);
+
+    // make sure servers can do rolling upgrade too
+    upgradeServer("server1", server1Port, locatorPort, startupExecution);
+    causeP2PTraffic(locatorPort);
+
+    upgradeServer("server2", server2Port, locatorPort, startupExecution);
+    causeP2PTraffic(locatorPort);
+  }
+
+  private void upgradeLocator(String name, int locatorPort, int locatorJmxPort,
+      int connectedLocatorPort,
+      GfshExecution startupExecution) {
+    oldGfsh.stopLocator(startupExecution, name);
+    GfshScript
+        .of(startLocatorCommandWithConfig(name, locatorPort, locatorJmxPort, connectedLocatorPort))
+        .execute(currentGfsh, tempFolder.getRoot());
+  }
+
+  private void upgradeServer(String name, int serverPort, int locatorPort,
+      GfshExecution startupExecution) {
+    oldGfsh.stopServer(startupExecution, name);
+    GfshScript.of(startServerCommandWithConfig(name, serverPort, locatorPort))
```

Comment:
It will be nice to add verification to see old locator, server was successfully stopped and new locator and server started. I am thinking of case where the old locator and servers were not stopped. And in the verifyMembers we are only looking at any servers or locators (not version specific).

Reply:
We'll consider these improvement suggestions in a future PR.